File: games/tbamud/commands/info.py
"""Information commands — time, weather, consider, examine, score, who, inventory, etc.

Overrides core engine commands with original tbaMUD-style output (한글화).
Based on act.informative.c from CircleMUD/tbaMUD.
"""

from __future__ import annotations

from collections import Counter
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from core.engine import Engine
    from core.session import Session


def register(engine: Engine) -> None:
    engine.register_command("time", do_time, korean="시간")
    engine.register_command("weather", do_weather, korean="날씨")
    engine.register_command("examine", do_examine, korean="조사")
    engine.register_command("consider", do_consider, korean="평가")
    engine.register_command("where", do_where, korean="어디")
    # Override core commands with original tbaMUD format
    engine.register_command("score", do_score, korean="점수")
    engine.register_command("who", do_who, korean="누구")
    engine.register_command("inventory", do_inventory, korean="소지품")
    engine.register_command("i", do_inventory)
    engine.register_command("equipment", do_equipment, korean="장비")
    engine.register_command("eq", do_equipment)
    engine.register_command("exits", do_exits, korean="출구")


async def do_time(session: Session, args: str) -> None:
    await session.send_line("현재 시각을 알 수 없습니다. (구현 예정)")


async def do_weather(session: Session, args: str) -> None:
    await session.send_line("날씨는 맑습니다. (구현 예정)")


async def do_examine(session: Session, args: str) -> None:
    if not args:
        await session.send_line("무엇을 조사하시겠습니까?")
        return
    # For now, same as look
    await session.engine.do_look(session, args)


async def do_consider(session: Session, args: str) -> None:
    if not args:
        await session.send_line("누구를 평가하시겠습니까?")
        return
    char = session.character
    if not char:
        return
    room = session.engine.world.get_room(char.room_vnum)
    if not room:
        return
    target_kw = args.strip().lower()
    for mob in room.characters:
        if mob is char:
            continue
        if mob.is_npc and target_kw in mob.proto.keywords.lower():
            diff = mob.level - char.level
            if diff <= -10:
                msg = "이제 눈을 감고도 이길 수 있습니다."
            elif diff <= -5:
                msg = "쉬운 상대입니다."
            elif diff <= -2:
                msg = "어렵지 않은 상대입니다."
            elif diff <= 2:
                msg = "꽤 대등한 상대입니다."
            elif diff <= 5:
                msg = "조금 힘든 싸움이 될 것 같습니다."
            elif diff <= 10:
                msg = "매우 위험한 상대입니다!"
            else:
                msg = "자살 행위입니다!!!"
            await session.send_line(msg)
            return
    await session.send_line("그런 대상을 찾을 수 없습니다.")


async def do_where(session: Session, args: str) -> None:
    """Show location of players or named mobs in the same zone."""
    char = session.character
    if not char:
        return
    room = session.engine.world.get_room(char.room_vnum)
    if not room:
        return
    zone = room.proto.zone_vnum

    if not args:
        # Show all players in same zone
        lines = ["{bright_cyan}-- 주변 플레이어 --{reset}"]
        found = False
        for rm in session.engine.world.rooms.values():
            if rm.proto.zone_vnum != zone:
                continue
            for ch in rm.characters:
                if not ch.is_npc and ch is not char:
                    lines.append(f"  {ch.name} — {rm.proto.name}")
                    found = True
        if not found:
            lines.append("  주변에 다른 플레이어가 없습니다.")
        await session.send_line("\r\n".join(lines))
    else:
        target_kw = args.strip().lower()
        lines = [f"{{bright_cyan}}-- '{args}' 탐색 결과 --{{reset}}"]
        found = False
        for rm in session.engine.world.rooms.values():
            if rm.proto.zone_vnum != zone:
                continue
            for ch in rm.characters:
                if target_kw in ch.proto.keywords.lower() or (ch.player_name and target_kw in ch.player_name.lower()):
                    lines.append(f"  {ch.name} — {rm.proto.name}")
                    found = True
        if not found:
            lines.append("  찾을 수 없습니다.")
        await session.send_line("\r\n".join(lines))


# ── Original tbaMUD format overrides ─────────────────────────────

# Position names (from tbaMUD act.informative.c)
_POS_NAMES = {
    0: "죽어 있습니다",
    1: "빈사 상태입니다",
    2: "의식불명입니다",
    3: "기절해 있습니다",
    4: "잠들어 있습니다",
    5: "쉬고 있습니다",
    6: "앉아 있습니다",
    7: "전투 중입니다",
    8: "서 있습니다",
}

# tbaMUD 18 wear positions (from constants.c wear_where[])
_WEAR_WHERE = [
    "<머리 위에>        ",  # 0 WEAR_LIGHT
    "<왼손 손가락에>    ",  # 1 WEAR_FINGER_R
    "<오른손 손가락에>  ",  # 2 WEAR_FINGER_L
    "<목에>             ",  # 3 WEAR_NECK_1
    "<목에>             ",  # 4 WEAR_NECK_2
    "<몸통에>           ",  # 5 WEAR_BODY
    "<머리에>           ",  # 6 WEAR_HEAD
    "<다리에>           ",  # 7 WEAR_LEGS
    "<발에>             ",  # 8 WEAR_FEET
    "<손에>             ",  # 9 WEAR_HANDS
    "<팔에>             ",  # 10 WEAR_ARMS
    "<방패로>           ",  # 11 WEAR_SHIELD
    "<몸 주위에>        ",  # 12 WEAR_ABOUT
    "<허리에>           ",  # 13 WEAR_WAIST
    "<왼쪽 손목에>      ",  # 14 WEAR_WRIST_R
    "<오른쪽 손목에>    ",  # 15 WEAR_WRIST_L
    "<오른손에 들고>    ",  # 16 WEAR_WIELD
    "<왼손에 들고>      ",  # 17 WEAR_HOLD
]


def _obj_modifiers(obj) -> str:
    """Return modifier string for object display (original show_obj_modifiers)."""
    mods = []
    extra_flags = getattr(obj.proto, "extra_flags", [])
    if 1 in extra_flags:  # ITEM_GLOW
        mods.append("..빛나고 있습니다!")
    if 2 in extra_flags:  # ITEM_HUM
        mods.append("..윙윙거리는 소리가 납니다!")
    if 3 in extra_flags:  # ITEM_INVISIBLE
        mods.append("(투명)")
    return " ".join(mods)


async def do_score(session: Session, args: str) -> None:
    """tbaMUD original do_score format (한글화).

    Based on act.informative.c ACMD(do_score).
    """
    from games.tbamud.level import CLASS_NAMES, exp_to_next

    char = session.character
    if not char:
        return

    cls_name = CLASS_NAMES.get(char.class_id, "모험가")
    sex_name = ["중성", "남성", "여성"][session.player_data.get("sex", 0)]
    age = session.player_data.get("age", 17)
    title = session.player_data.get("title", "")

    # Compute AC from base + equipment affects
    ac = 100
    for obj in char.equipment.values():
        for aff in getattr(obj.proto, "affects", []):
            loc = aff.get("location", "")
            if loc == "ARMOR" or loc == "AC":
                ac += aff.get("modifier", 0)

    alignment = getattr(char, "alignment", getattr(char.proto, "alignment", 0))

    lines = []

    # Age
    lines.append(f"당신은 {age}살입니다.")

    # HP / Mana / Move
    lines.append(
        f"체력 {{green}}{char.hp}({char.max_hp}){{reset}}, "
        f"마나 {{cyan}}{char.mana}({char.max_mana}){{reset}}, "
        f"이동력 {char.move}({char.max_move})입니다."
    )

    # AC / Alignment
    lines.append(f"방어도 {ac // 10}/10, 성향 {alignment}입니다.")

    # Experience / Gold
    lines.append(
        f"경험치 {{yellow}}{char.experience:,}{{reset}}, "
        f"골드 {{yellow}}{char.gold:,}{{reset}}입니다."
    )

    # Next level (only for mortals)
    if char.level < 31:
        needed = exp_to_next(char)
        lines.append(f"다음 레벨까지 {{yellow}}{needed:,}{{reset}} 경험치가 필요합니다.")

    # Stats
    lines.append(
        f"힘: {char.str}  민첩: {char.dex}  체력: {char.con}  "
        f"지능: {char.intel}  지혜: {char.wis}  매력: {char.cha}"
    )

    # Hitroll / Damroll
    lines.append(f"히트롤: {char.hitroll}  댐롤: {char.damroll}")

    # Rank line
    rank_line = f"당신은 레벨 {char.level} {cls_name} {char.name}"
    if title:
        rank_line += f" {title}"
    rank_line += "입니다."
    lines.append(rank_line)

    # Position
    if char.position == 7 and char.fighting:
        lines.append(f"현재 {char.fighting.name}과(와) 전투 중입니다!")
    else:
        pos_msg = _POS_NAMES.get(char.position, "서 있습니다")
        lines.append(f"현재 {pos_msg}.")

    # Active affects
    affect_names = []
    for aff in char.affects:
        name = aff.get("name", aff.get("spell", ""))
        if name:
            dur = aff.get("duration", 0)
            affect_names.append(f"{name}({dur})")
    if affect_names:
        lines.append(f"활성 효과: {', '.join(affect_names)}")

    await session.send_line("\r\n".join(lines))


async def do_who(session: Session, args: str) -> None:
    """tbaMUD original do_who format — Immortals/Mortals split.

    Based on act.informative.c ACMD(do_who).
    """
    from games.tbamud.level import CLASS_NAMES

    engine = session.engine

    immortals = []
    mortals = []

    for name, s in engine.players.items():
        if not s.character:
            continue
        c = s.character
        cls_id = s.player_data.get("class_id", 0)
        cls_name = CLASS_NAMES.get(cls_id, "모험")[:3]
        lvl = s.player_data.get("level", 1)
        title = s.player_data.get("title", "")
        display = f"[{lvl:3d} {cls_name:3s}] {c.name}"
        if title:
            display += f" {title}"
        # Flags
        flags = []
        if c.position == 8:
            pass
        elif c.position == 4:
            flags.append("잠듦")
        elif c.position == 5:
            flags.append("휴식")
        elif c.position == 7:
            flags.append("전투")
        if flags:
            display += f" ({', '.join(flags)})"

        if lvl >= 31:
            immortals.append(display)
        else:
            mortals.append(display)

    lines = []
    if immortals:
        lines.append("{yellow}신들{reset}")
        lines.append("{yellow}────{reset}")
        lines.extend(f"  {{yellow}}{im}{{reset}}" for im in immortals)
        lines.append("")

    if mortals:
        lines.append("모험가들")
        lines.append("────────")
        lines.extend(f"  {m}" for m in mortals)
        lines.append("")

    total = len(immortals) + len(mortals)
    lines.append(f"{total}명의 모험가가 접속 중입니다.")

    await session.send_line("\r\n".join(lines))


async def do_inventory(session: Session, args: str) -> None:
    """tbaMUD original do_inventory — grouped items with count.

    Based on act.informative.c ACMD(do_inventory).
    """
    char = session.character
    if not char:
        return

    await session.send_line("소지품:")
    if not char.inventory:
        await session.send_line("  아무것도 없습니다.")
        return

    # Group by short_description
    counts: Counter[str] = Counter()
    name_map: dict[str, str] = {}
    for obj in char.inventory:
        desc = obj.proto.short_desc
        counts[desc] += 1
        if desc not in name_map:
            mods = _obj_modifiers(obj)
            name_map[desc] = f"{desc} {mods}".rstrip() if mods else desc

    for desc, count in counts.items():
        display = name_map[desc]
        if count > 1:
            await session.send_line(f"  ({count}) {display}")
        else:
            await session.send_line(f"  {display}")


async def do_equipment(session: Session, args: str) -> None:
    """tbaMUD original do_equipment — wear_where labels + 18 slots.

    Based on act.informative.c ACMD(do_equipment).
    """
    char = session.character
    if not char:
        return

    await session.send_line("착용 장비:")
    found = False
    for pos in range(len(_WEAR_WHERE)):
        obj = char.equipment.get(pos)
        if obj:
            mods = _obj_modifiers(obj)
            name = obj.proto.short_desc
            if mods:
                name = f"{name} {mods}"
            await session.send_line(f"  {_WEAR_WHERE[pos]}{name}")
            found = True

    if not found:
        await session.send_line("  아무것도 착용하고 있지 않습니다.")


async def do_exits(session: Session, args: str) -> None:
    """tbaMUD original do_exits — direction labels with room names.

    Based on act.informative.c ACMD(do_exits).
    """
    from core.engine import DIR_NAMES_KR

    char = session.character
    if not char:
        return
    room = session.engine.world.get_room(char.room_vnum)
    if not room:
        return

    lines = ["확인 가능한 출구:"]
    found = False
    for ex in room.proto.exits:
        if ex.direction >= 6:
            continue
        dir_name = DIR_NAMES_KR[ex.direction]
        if room.has_door(ex.direction):
            if room.is_door_closed(ex.direction):
                kw = ex.keywords.strip() if ex.keywords else "문"
                lines.append(f"  {dir_name:4s} - {kw}이(가) 닫혀 있습니다.")
                found = True
                continue

        dest = session.engine.world.get_room(ex.to_room)
        dest_name = dest.name if dest else "알 수 없음"
        lines.append(f"  {dir_name:4s} - {dest_name}")
        found = True

    if not found:
        lines.append("  없습니다.")

    await session.send_line("\r\n".join(lines))

File: games/tbamud/lua/commands/info.lua
-- info.lua — tbaMUD information commands (overrides common score/who/etc.)
-- Based on act.informative.c from CircleMUD/tbaMUD, 한글화.

-- Position names (from tbaMUD act.informative.c)
local POS_NAMES = {
    [0] = "죽어 있습니다",
    [1] = "빈사 상태입니다",
    [2] = "의식불명입니다",
    [3] = "기절해 있습니다",
    [4] = "잠들어 있습니다",
    [5] = "쉬고 있습니다",
    [6] = "앉아 있습니다",
    [7] = "전투 중입니다",
    [8] = "서 있습니다",
}

-- Wear position labels (18 slots, from constants.c wear_where[])
local WEAR_WHERE = {
    [0]  = "<머리 위에>        ",
    [1]  = "<왼손 손가락에>    ",
    [2]  = "<오른손 손가락에>  ",
    [3]  = "<목에>             ",
    [4]  = "<목에>             ",
    [5]  = "<몸통에>           ",
    [6]  = "<머리에>           ",
    [7]  = "<다리에>           ",
    [8]  = "<발에>             ",
    [9]  = "<손에>             ",
    [10] = "<팔에>             ",
    [11] = "<방패로>           ",
    [12] = "<몸 주위에>        ",
    [13] = "<허리에>           ",
    [14] = "<왼쪽 손목에>      ",
    [15] = "<오른쪽 손목에>    ",
    [16] = "<오른손에 들고>    ",
    [17] = "<왼손에 들고>      ",
}

-- Class names
local CLASS_NAMES = {
    [0] = "마법사", [1] = "성직자", [2] = "도적", [3] = "전사",
}

-- Experience tables (per class, per level)
local EXP_TABLE = {
    [0] = {
        [0]=0, [1]=1, [2]=2500, [3]=5000, [4]=10000, [5]=20000, [6]=40000,
        [7]=60000, [8]=90000, [9]=135000, [10]=250000, [11]=375000, [12]=750000,
        [13]=1125000, [14]=1500000, [15]=1875000, [16]=2250000, [17]=2625000,
        [18]=3000000, [19]=3375000, [20]=3750000, [21]=4000000, [22]=4300000,
        [23]=4600000, [24]=4900000, [25]=5200000, [26]=5500000, [27]=5950000,
        [28]=6400000, [29]=6850000, [30]=7400000, [31]=8000000,
    },
    [1] = {
        [0]=0, [1]=1, [2]=1500, [3]=3000, [4]=6000, [5]=13000, [6]=27500,
        [7]=55000, [8]=110000, [9]=225000, [10]=450000, [11]=675000, [12]=900000,
        [13]=1125000, [14]=1350000, [15]=1575000, [16]=1800000, [17]=2100000,
        [18]=2400000, [19]=2700000, [20]=3000000, [21]=3250000, [22]=3500000,
        [23]=3800000, [24]=4100000, [25]=4400000, [26]=4800000, [27]=5200000,
        [28]=5600000, [29]=6000000, [30]=6400000, [31]=7000000,
    },
    [2] = {
        [0]=0, [1]=1, [2]=1250, [3]=2500, [4]=5000, [5]=10000, [6]=20000,
        [7]=40000, [8]=70000, [9]=110000, [10]=160000, [11]=220000, [12]=440000,
        [13]=660000, [14]=880000, [15]=1100000, [16]=1500000, [17]=2000000,
        [18]=2500000, [19]=3000000, [20]=3500000, [21]=3650000, [22]=3800000,
        [23]=4100000, [24]=4400000, [25]=4700000, [26]=5100000, [27]=5500000,
        [28]=5900000, [29]=6300000, [30]=6650000, [31]=7000000,
    },
    [3] = {
        [0]=0, [1]=1, [2]=2000, [3]=4000, [4]=8000, [5]=16000, [6]=32000,
        [7]=64000, [8]=125000, [9]=250000, [10]=500000, [11]=750000, [12]=1000000,
        [13]=1250000, [14]=1500000, [15]=1850000, [16]=2200000, [17]=2550000,
        [18]=2900000, [19]=3250000, [20]=3600000, [21]=3900000, [22]=4200000,
        [23]=4500000, [24]=4800000, [25]=5150000, [26]=5500000, [27]=5950000,
        [28]=6400000, [29]=6850000, [30]=7400000, [31]=8000000,
    },
}

local MAX_LEVEL = 34

local function get_class_name(class_id)
    return CLASS_NAMES[class_id] or "모험가"
end

local function exp_for_level(class_id, level)
    local tbl = EXP_TABLE[class_id] or EXP_TABLE[0]
    level = math.min(level, 31)
    return tbl[level] or tbl[31] or 8000000
end

local function exp_to_next(ch)
    if ch.level >= 31 then return 0 end
    local needed = exp_for_level(ch.class_id, ch.level + 1)
    return math.max(0, needed - ch.experience)
end

local function obj_modifiers(obj)
    local mods = {}
    local flags = obj.proto.flags
    if not flags then return "" end
    -- Iterate flags list
    local ok, len = pcall(function() return #flags end)
    if not ok then return "" end
    for i = 0, len - 1 do
        local ok2, flag = pcall(function() return flags[i] end)
        if ok2 then
            if flag == 1 then table.insert(mods, "..빛나고 있습니다!") end
            if flag == 2 then table.insert(mods, "..윙윙거리는 소리가 납니다!") end
            if flag == 3 then table.insert(mods, "(투명)") end
        end
    end
    return table.concat(mods, " ")
end

local function format_number(n)
    local s = tostring(math.floor(n))
    local result = ""
    local len = #s
    for i = 1, len do
        if i > 1 and (len - i + 1) % 3 == 0 then
            result = result .. ","
        end
        result = result .. s:sub(i, i)
    end
    return result
end

-- ── time ────────────────────────────────────────────────────────

register_command("time", function(ctx, args)
    ctx:send("현재 시각을 알 수 없습니다. (구현 예정)")
end, "시간")

-- ── weather ─────────────────────────────────────────────────────

register_command("weather", function(ctx, args)
    ctx:send("날씨는 맑습니다. (구현 예정)")
end, "날씨")

-- ── examine ─────────────────────────────────────────────────────

register_command("examine", function(ctx, args)
    if not args or args == "" then
        ctx:send("무엇을 조사하시겠습니까?")
        return
    end
    ctx:call_command("look", args)
end, "조사")

-- ── consider ────────────────────────────────────────────────────

register_command("consider", function(ctx, args)
    if not args or args == "" then
        ctx:send("누구를 평가하시겠습니까?")
        return
    end
    local ch = ctx.char
    if not ch then return end
    local target = ctx:find_char(args)
    if not target then
        ctx:send("그런 대상을 찾을 수 없습니다.")
        return
    end
    local diff = target.level - ch.level
    local msg
    if diff <= -10 then
        msg = "이제 눈을 감고도 이길 수 있습니다."
    elseif diff <= -5 then
        msg = "쉬운 상대입니다."
    elseif diff <= -2 then
        msg = "어렵지 않은 상대입니다."
    elseif diff <= 2 then
        msg = "꽤 대등한 상대입니다."
    elseif diff <= 5 then
        msg = "조금 힘든 싸움이 될 것 같습니다."
    elseif diff <= 10 then
        msg = "매우 위험한 상대입니다!"
    else
        msg = "자살 행위입니다!!!"
    end
    ctx:send(msg)
end, "평가")

-- ── where ───────────────────────────────────────────────────────

register_command("where", function(ctx, args)
    local ch = ctx.char
    if not ch then return end

    if not args or args == "" then
        local results = ctx:get_zone_chars()
        ctx:send("{bright_cyan}-- 주변 플레이어 --{reset}")
        local found = false
        for i = 1, #results do
            local r = results[i]
            if not r.char.is_npc then
                ctx:send("  " .. r.char.name .. " — " .. r.room_name)
                found = true
            end
        end
        if not found then
            ctx:send("  주변에 다른 플레이어가 없습니다.")
        end
    else
        local results = ctx:get_zone_chars(nil, args)
        ctx:send("{bright_cyan}-- '" .. args .. "' 탐색 결과 --{reset}")
        if #results == 0 then
            ctx:send("  찾을 수 없습니다.")
        else
            for i = 1, #results do
                local r = results[i]
                ctx:send("  " .. r.char.name .. " — " .. r.room_name)
            end
        end
    end
end, "어디")

-- ── score (tbaMUD original format) ──────────────────────────────

register_command("score", function(ctx, args)
    local ch = ctx.char
    if not ch then return end

    local cls_name = get_class_name(ch.class_id)
    local sex_name = "중성"
    local sex = ctx:get_player_data("sex") or 0
    if sex == 1 then sex_name = "남성"
    elseif sex == 2 then sex_name = "여성" end
    local age = ctx:get_player_data("age") or 17
    local title = ctx:get_player_data("title") or ""

    -- Compute AC from base + equipment
    local ac = 100
    local equip = ctx:get_equipment()
    for i = 1, #equip do
        local obj = equip[i].obj
        local affects = obj.proto.affects
        if affects then
            local ok, len = pcall(function() return #affects end)
            if ok then
                for j = 0, len - 1 do
                    local aff_ok, aff = pcall(function() return affects[j] end)
                    if aff_ok and aff then
                        local loc = aff.location or ""
                        if loc == "ARMOR" or loc == "AC" then
                            ac = ac + (aff.modifier or 0)
                        end
                    end
                end
            end
        end
    end

    local alignment = ch.proto.alignment or 0

    -- Age
    ctx:send("당신은 " .. age .. "살입니다.")

    -- HP / Mana / Move
    ctx:send("체력 {green}" .. ch.hp .. "(" .. ch.max_hp .. "){reset}, " ..
             "마나 {cyan}" .. ch.mana .. "(" .. ch.max_mana .. "){reset}, " ..
             "이동력 " .. ch.move .. "(" .. ch.max_move .. ")입니다.")

    -- AC / Alignment
    ctx:send("방어도 " .. math.floor(ac / 10) .. "/10, 성향 " .. alignment .. "입니다.")

    -- Experience / Gold
    ctx:send("경험치 {yellow}" .. format_number(ch.experience) .. "{reset}, " ..
             "골드 {yellow}" .. format_number(ch.gold) .. "{reset}입니다.")

    -- Next level
    if ch.level < 31 then
        local needed = exp_to_next(ch)
        ctx:send("다음 레벨까지 {yellow}" .. format_number(needed) .. "{reset} 경험치가 필요합니다.")
    end

    -- Stats
    local stats = ch.stats or {}
    local function gs(k) local ok, v = pcall(function() return stats[k] end); return ok and v or 0 end
    ctx:send("힘: " .. gs("str") .. "  민첩: " .. gs("dex") .. "  체력: " .. gs("con") ..
             "  지능: " .. gs("int") .. "  지혜: " .. gs("wis") .. "  매력: " .. gs("cha"))

    -- Hitroll / Damroll
    ctx:send("히트롤: " .. ch.hitroll .. "  댐롤: " .. ch.damroll)

    -- Rank line
    local rank = "당신은 레벨 " .. ch.level .. " " .. cls_name .. " " .. ch.name
    if title and title ~= "" then
        rank = rank .. " " .. title
    end
    rank = rank .. "입니다."
    ctx:send(rank)

    -- Position
    if ch.position == 7 and ch.fighting then
        ctx:send("현재 " .. ch.fighting.name .. "과(와) 전투 중입니다!")
    else
        local pos_msg = POS_NAMES[ch.position] or "서 있습니다"
        ctx:send("현재 " .. pos_msg .. ".")
    end

    -- Active affects
    local aff_names = {}
    local affects = ch.affects
    if affects then
        local ok, len = pcall(function() return #affects end)
        if ok then
            for i = 0, len - 1 do
                local aff_ok, aff = pcall(function() return affects[i] end)
                if aff_ok and aff then
                    local name = aff.name or aff.spell or ""
                    if name ~= "" then
                        local dur = aff.duration or 0
                        table.insert(aff_names, name .. "(" .. dur .. ")")
                    end
                end
            end
        end
    end
    if #aff_names > 0 then
        ctx:send("활성 효과: " .. table.concat(aff_names, ", "))
    end
end, "점수")

-- ── who (tbaMUD original format — Immortals/Mortals split) ──────

register_command("who", function(ctx, args)
    local players = ctx:get_players()
    local immortals = {}
    local mortals = {}

    for i = 1, #players do
        local p = players[i]
        local cls_name = get_class_name(p.class_id)
        if #cls_name > 3 then cls_name = cls_name:sub(1, 3) end
        local level = p.level
        local display = "[" .. string.format("%3d", level) .. " " ..
                        string.format("%-3s", cls_name) .. "] " .. p.name

        -- Flags
        local flags = {}
        if p.position == 4 then table.insert(flags, "잠듦") end
        if p.position == 5 then table.insert(flags, "휴식") end
        if p.position == 7 then table.insert(flags, "전투") end
        if #flags > 0 then
            display = display .. " (" .. table.concat(flags, ", ") .. ")"
        end

        if level >= 31 then
            table.insert(immortals, display)
        else
            table.insert(mortals, display)
        end
    end

    local lines = {}
    if #immortals > 0 then
        table.insert(lines, "{yellow}신들{reset}")
        table.insert(lines, "{yellow}────{reset}")
        for _, im in ipairs(immortals) do
            table.insert(lines, "  {yellow}" .. im .. "{reset}")
        end
        table.insert(lines, "")
    end
    if #mortals > 0 then
        table.insert(lines, "모험가들")
        table.insert(lines, "────────")
        for _, m in ipairs(mortals) do
            table.insert(lines, "  " .. m)
        end
        table.insert(lines, "")
    end
    local total = #immortals + #mortals
    table.insert(lines, total .. "명의 모험가가 접속 중입니다.")
    ctx:send(table.concat(lines, "\r\n"))
end, "누구")

-- ── inventory (tbaMUD grouped items) ────────────────────────────

register_command("inventory", function(ctx, args)
    local ch = ctx.char
    if not ch then return end
    local inv = ctx:get_inventory()

    ctx:send("소지품:")
    if #inv == 0 then
        ctx:send("  아무것도 없습니다.")
        return
    end

    -- Group by short_description
    local counts = {}
    local order = {}
    local display_map = {}
    for i = 1, #inv do
        local obj = inv[i]
        local desc = obj.proto.short_desc
        if not counts[desc] then
            counts[desc] = 0
            table.insert(order, desc)
            local mods = obj_modifiers(obj)
            if mods ~= "" then
                display_map[desc] = desc .. " " .. mods
            else
                display_map[desc] = desc
            end
        end
        counts[desc] = counts[desc] + 1
    end

    for _, desc in ipairs(order) do
        local display = display_map[desc]
        if counts[desc] > 1 then
            ctx:send("  (" .. counts[desc] .. ") " .. display)
        else
            ctx:send("  " .. display)
        end
    end
end, "소지품")

-- i → inventory alias
register_command("i", function(ctx, args)
    ctx:call_command("inventory", args or "")
end)

-- ── equipment (tbaMUD 18-slot display) ──────────────────────────

register_command("equipment", function(ctx, args)
    local ch = ctx.char
    if not ch then return end
    local equip = ctx:get_equipment()

    ctx:send("착용 장비:")
    if #equip == 0 then
        ctx:send("  아무것도 착용하고 있지 않습니다.")
        return
    end

    for i = 1, #equip do
        local slot = equip[i].slot
        local obj = equip[i].obj
        local label = WEAR_WHERE[slot] or "<어딘가에>          "
        local mods = obj_modifiers(obj)
        local name = obj.proto.short_desc
        if mods ~= "" then name = name .. " " .. mods end
        ctx:send("  " .. label .. name)
    end
end, "장비")

-- eq → equipment alias
register_command("eq", function(ctx, args)
    ctx:call_command("equipment", args or "")
end)

-- ── exits (tbaMUD format with room names) ───────────────────────

register_command("exits", function(ctx, args)
    local ch = ctx.char
    if not ch then return end
    local room = ctx:get_room()
    if not room then return end

    local exits = ctx:get_exits()
    local lines = {"확인 가능한 출구:"}
    local found = false

    for i = 1, #exits do
        local ex = exits[i]
        if ex.direction < 6 then
            local dir_name = DIR_NAMES[ex.direction + 1]
            if room:has_door(ex.direction) and room:is_door_closed(ex.direction) then
                local kw = ex.keywords or "문"
                if kw == "" then kw = "문" end
                kw = kw:match("^%S+") or kw
                table.insert(lines, "  " .. string.format("%-4s", dir_name) ..
                             " - " .. kw .. "이(가) 닫혀 있습니다.")
            else
                local dest = ctx:get_room(ex.to_room)
                local dest_name = "알 수 없음"
                if dest then dest_name = dest.name end
                table.insert(lines, "  " .. string.format("%-4s", dir_name) ..
                             " - " .. dest_name)
            end
            found = true
        end
    end

    if not found then
        table.insert(lines, "  없습니다.")
    end
    ctx:send(table.concat(lines, "\r\n"))
end, "출구")
